numpy/lib/function_base.py

```diff
@@ -4014,122 +4135,234 @@ def _quantile_is_valid(q):
+    values_count = arr.shape[axis]
+    if values_count == 0:
+        return np.NAN
+    if values_count == 1:
+        result = np.take(arr, indices=0, axis=axis)
+        if np.issubdtype(arr.dtype, np.inexact) and np.isnan(result):
+            warnings.warn("All-NaN slice encountered", RuntimeWarning,
+                          stacklevel=4)
+        return np.broadcast_to(result, (quantiles.size,) + result.shape)
     # The dimensions of `q` are prepended to the output shape, so we need the
     # axis being sampled from `ap` to be first.
-    ap = np.moveaxis(ap, axis, 0)
-    del axis
-
-    if np.issubdtype(indices.dtype, np.integer):
-        # take the points along axis
-
-        if np.issubdtype(a.dtype, np.inexact):
-            # may contain nan, which would sort to the end
-            ap.partition(concatenate((indices.ravel(), [-1])), axis=0)
-            n = np.isnan(ap[-1])
-        else:
-            # cannot contain nan
-            ap.partition(indices.ravel(), axis=0)
-            n = np.array(False, dtype=bool)
-
-        r = take(ap, indices, axis=0, out=out)
-
+    arr = np.moveaxis(arr, axis, destination=0)
+    valid_values_count = np.broadcast_to(not_scalar(values_count),
+                                         arr.shape[1:])
+    # We need at least two values to do an interpolation
+    too_few_values = valid_values_count < 2
+    if too_few_values.any():
+        valid_values_count[too_few_values] = np.NaN
+    # --- Computation of indexes
+    # Add axis for quantiles
+    valid_values_count = valid_values_count[..., np.newaxis]
+    # Index where to find the value in the sorted array.
+    # Virtual because it is a floating point value, not an valid index.
+    # The nearest neighbours are used for interpolation
+    virtual_indexes = np.where(
+        valid_values_count == 0,
+        0.0,
+        interpolation.virtual_index_fun(valid_values_count, quantiles),
+    )
+    # After sort, slices having NaNs will have for last element a NaN
+    previous_indexes = np.floor(virtual_indexes)
+    next_indexes = previous_indexes + 1
+    # When indexes is above max index, take the max value of the array
+    indexes_above_bounds = virtual_indexes >= valid_values_count - 1
+    if indexes_above_bounds.any():
+        previous_indexes[indexes_above_bounds] = -1
+        next_indexes[indexes_above_bounds] = -1
+    # When indexes is below min index, take the min value of the array
+    indexes_below_bounds = virtual_indexes < 0
+    if indexes_below_bounds.any():
+        previous_indexes[indexes_below_bounds] = 0
+        next_indexes[indexes_below_bounds] = 0
+    if np.issubdtype(arr.dtype, np.inexact):
+        virtual_indexes_nans = np.isnan(virtual_indexes)
+        if virtual_indexes_nans.any():
+            previous_indexes[virtual_indexes_nans] = -1
+            next_indexes[virtual_indexes_nans] = -1
+    previous_indexes = previous_indexes.astype(int)
+    next_indexes = next_indexes.astype(int)
```

Comment:
Speaking of splitting things into smaller functions: the generation of `next_indexes` and `previous_indexes` sounds like a prime candidate for their own dedicated function.

Reply:
Done